logbook-okhttp2/src/main/java/org/zalando/logbook/okhttp2/RemoteResponse.java

```diff
@@ -12,20 +14,124 @@
 import java.util.Locale;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static com.squareup.okhttp.ResponseBody.create;
 import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.util.Objects.requireNonNull;
+import static org.zalando.fauxpas.FauxPas.throwingUnaryOperator;
 
+@AllArgsConstructor
 final class RemoteResponse implements HttpResponse {
 
-    private Response response;
-    private byte[] body;
+    private interface State {
+
+        default State with() {
+            return this;
+        }
+
+        default State without() {
+            return this;
+        }
+
+        default State buffer() throws IOException {
+            return this;
+        }
+
+        Response getResponse();
+
+        default byte[] getBody() {
+            return new byte[0];
+        }
 
-    RemoteResponse(final Response response) {
-        this.response = response;
     }
 
+    private abstract class AbstractState implements State {
+
+        @Override
+        public Response getResponse() {
+            return response;
+        }
+
+    }
+
+    private final class Unbuffered extends AbstractState {
+
+        @Override
+        public State with() {
+            return new Offering();
+        }
+
+    }
+
+    private final class Offering extends AbstractState {
+
+        @Override
+        public State without() {
+            return new Unbuffered();
+        }
+
+        @Override
+        public State buffer() throws IOException {
+            final ResponseBody entity = requireNonNull(response.body(),
+                    "Body is never null for normal responses");
+
+            if (entity.contentLength() == 0L) {
+                return new Passing();
+            } else {
+                final byte[] body = entity.bytes();
+
+                final Response copy = response.newBuilder()
+                        .body(create(entity.contentType(), body))
+                        .build();
+
+                return new Buffering(copy, body);
+            }
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Buffering implements State {
+
+        @Getter
+        private final Response response;
+        private final byte[] body;
+
+        @Override
+        public State without() {
+            return new Ignoring(response, body);
+        }
+
+        @Override
+        public byte[] getBody() {
+            return body;
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static final class Ignoring implements State {
+
+        @Getter
+        private final Response response;
+        private final byte[] body;
+
+        @Override
+        public State with() {
+            return new Buffering(response, body);
+        }
+
+    }
+
+    private final class Passing extends AbstractState {
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)